### ISSUE TITLE
Kubenetes deployed Tasks which contain Spring Batch do not correlate status in the same execution database entry.

### ISSUE THREAD

**Issue by AdVirMachina** [spring-cloud/spring-cloud-dataflow#3628]
As detailed in #3536, SCDF does not correlate status correctly when Kubernetes is configured to launch Spring Batch Docker container in entrypoint style of SHELL.

I believe 3536 was closed incorrectly. Someone else jumped on that thread with a similar problem which was solved, but not the original report. I used the BILLRUN sample task from the Spring Cloud DataFlow batch samples. I did not do anything to that code. I was asked to upload my configuration files for MiniKube which I did. 

There is no reason to upload anything else than perhaps the docker file which created the container. Please find it and the script it executes attached.

[Dockerfile.gz](https://github.com/spring-cloud/spring-cloud-dataflow/files/3833481/Dockerfile.gz)
[entrypoint.sh.gz](https://github.com/spring-cloud/spring-cloud-dataflow/files/3833482/entrypoint.sh.gz)

**Steps to reproduce:**
Build the billing container using the Dockerfile and entrypoint.sh.
Store the billing container in the Docker repository within the Kubernetes cluster.
Modify the service-config.yaml file to have this entry:
```
          task:
            platform:
              kubernetes:
                accounts:
                  default:
                    entryPointStyle: shell
```
Load this updated config into Kubernetes.
Return to SCDF.
Configure SCDF to define an Application loading the Billing Container as a TASK.
Create a composed task which executes billrun and execute it.
Under Executions, you should see the deployment (per #3536's screenshots) and then when billrun is complete, the actual Spring Batch execution under a different (usually the next available) Execution ID.

What the Spring Batch job does inside of the container is not important. This will always be the manifested behavior from Spring Batch. If you create a dirt simple Spring Batch job that runs steps which simply write to System.out, you get the same thing.

**Comment by tzolov** [comment]
@AdVirMachina, find below two workarounds that should work with your example but handle the input properties and arguments differently. It will help if you can share reasons to go for the SHELL form execution? 
Next I will elaborate on the differences between the `SHELL` and `EXEC` forms as defined in the Docker specification and the implication this imposes to Spring Cloud Tasks and Spring Batch jobs.

#### Workaround 1
* Change the entrypoint style to [boot](https://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#_entry_point_style) instead of `shell`. Note that you may want to set this as deployer instead of a global property. Setting it as a global property would apply for all Tasks run the SCDF.
* Launch the task with the following deployment property: `deployer.<your-registered-task-app-name>.kubernetes.containerCommand=/entrypoint.sh`. Replace the `<your-registered-task-app-name>` with the name of the task you have registered.

   Alternatively you can replace the `CMD /entrypoint.sh` (which is redundant when command line arguments are present) by a `ENTRYPOINT /entrypoint.sh` instead. The result is that the script is still run within a shell (e.g. `/bin/sh /entrypoint.sh ..`) but there is no need to set the `containerCommand `deployment property.

   **NOTE**: The boot entryPointStyle creates an environment variable called `SPRING_APPLICATION_JSON` that contains a JSON representation of all application properties. The command line arguments from the deployment request are set as container args: https://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#_entry_point_style
For example here are the billrun's container args and variables:
```
Args:
      --spring.cloud.data.flow.platformname=default
      --spring.cloud.task.executionid=18
Environment:
      SPRING_APPLICATION_JSON:        {"spring.datasource.username":"root","management.metrics.export.prometheus.enabled":"true","spring.datasource.url":"jdbc:mysql://10.104.87.227:3306/mysql","spring.datasource.driverClassName":"org.mariadb.jdbc.Driver","management.metrics.export.prometheus.rsocket.port":"7001","management.metrics.export.prometheus.rsocket.enabled":"true","spring.cloud.task.name":"billrun-original-shell","management.metrics.export.prometheus.rsocket.host":"prometheus-proxy","spring.datasource.password":"yourpassword"}
      SPRING_CLOUD_APPLICATION_GUID:   (v1:metadata.uid)
```
And the  billrun java process is invoked from within a shell: 
```
PID TTY      STAT   TIME COMMAND
  1 ?        Ss     0:00 /bin/sh /entrypoint.sh --spring.cloud.data.flow.platformname=default --spring.cloud.task.executionid=19
  7 ?        Sl     0:00 java -Xms64m -Xmx64m -jar /app.jar --spring.cloud.data.flow.platformname=default --spring.cloud.task.executionid=19
 22 pts/0    Ss     0:00 bash
 27 pts/0    R+     0:00 ps axw
```

#### Workaround 2
* Set the entrypoint style to `EXEC`. 
* in `Dockerfile` replace `CMD /entrypoint.sh` by `ENTRYPOINT /entrypoint.sh`. In this case the entyrpoint.sh script will be invoked within a shell but all command line arguments are passed to the script.

   **Note:** unlike workaround 1 (using the BOOT style) the input properties are passed as command line arguments instead of being squashed as fields in the `SPRING_APPLICATION_JSON` variable: 
```
    Args:
      --spring.datasource.username=root
      --management.metrics.export.prometheus.enabled=true
      --spring.datasource.url=jdbc:mysql://10.111.157.226:3306/mysql
      --spring.datasource.driverClassName=org.mariadb.jdbc.Driver
      --management.metrics.export.prometheus.rsocket.port=7001
      --management.metrics.export.prometheus.rsocket.enabled=true
      --spring.cloud.task.name=billrunshell2
      --management.metrics.export.prometheus.rsocket.host=prometheus-proxy
      --spring.datasource.password=yourpassword
      --spring.cloud.data.flow.platformname=default
      --spring.cloud.task.executionid=5

    Environment:
      SPRING_CLOUD_APPLICATION_GUID:   (v1:metadata.uid)
```
and the corresponding container process execution looks like this:
```
PID TTY      STAT   TIME COMMAND
    1 ?        Ss     0:00 /bin/sh /entrypoint.sh --spring.datasource.username=root --management.metrics.export.prometheus.enabled=true --spring.datasource.url=jdbc:mysql://10.111.157.226:3306/mysql --spring.datasource.driverClassName=org.mariadb.jdbc.Driver --management.metrics.export.prometheus.rsocket.port=7001 --management.metrics.export.prometheus.rsocket.enabled=true --spring.cloud.task.name=billrunshell2 --management.metrics.export.prometheus.rsocket.host=prometheus-proxy --spring.datasource.password=yourpassword --spring.cloud.data.flow.platformname=default --spring.cloud.task.executionid=5
    7 ?        Sl     0:00 java -Xms64m -Xmx64m -jar /app.jar --spring.datasource.username=root --management.metrics.export.prometheus.enabled=true --spring.datasource.url=jdbc:mysql://10.111.157.226:3306/mysql --spring.datasource.driverClassName=org.mariadb.jdbc.Driver --management.metrics.export.prometheus.rsocket.port=7001 --management.metrics.export.prometheus.rsocket.enabled=true --spring.cloud.task.name=billrunshell2 --management.metrics.export.prometheus.rsocket.host=prometheus-proxy --spring.datasource.password=yourpassword --spring.cloud.data.flow.platformname=default --spring.cloud.task.executionid=5
```
#### Explanation
The docker specification, states that the [SHELL form prevents any CMD or run command line arguments from being used](https://docs.docker.com/engine/reference/builder/#entrypoint). E.g. any additional command line arguments passed to the `./etnrypoint.sh` are ignored. On the other hand the `Spring Batch` specification and inherently `Spring Cloud Task` rely on command line arguments for configuring dynamic properties! Additionally SCDF adds two special command line arguments on Task launch `spring.cloud.data.flow.platformname` and `spring.cloud.task.executionid`.

The way SHELL form handles command line arguments (e.g. dropping them) makes it inappropriate for running Spring Cloud Task/Batch applications, because Task and Batch apps can accept command line args as parameters. This is not SCDF related constraint but incompatibility of the Docker SHELL and Spring Cloud Task/Batch semantics.

Let’s me demonstrate this with an example showing how the command line arguments are handled by the SHELL and EXEC forms:
* SHELL - run SHELL form by invoking the command shell directly:
```
ENTRYPOINT ["/bin/sh", "-c", "java -Xms64m -Xmx64m -jar /app.jar"]
```
Run the shell-form image with additional `--foo=bar` command line argument:
```
$ docker run springcloudtask/billrun:shellform --foo=bar &
$ docker exec -it $(docker ps -q) ps axw
PID TTY      STAT   TIME COMMAND
  1 ?        Ss     0:00 /bin/sh -c java -Xms64m -Xmx64m -jar /app.jar --foo=bar
  6 ?        Sl     0:00 java -Xms64m -Xmx64m -jar /app.jar
 21 pts/0    Rs+    0:00 ps axw
```
Note that by the time the shell process (PID 1) start the java process (PID 6) the command line argument (--foo=bar) is dropped!

* EXEC - define an EXEC form entry point:
```
ENTRYPOINT ["java", "-Xms64m", "-Xmx64m", "-jar", "/app.jar"]
```
Run the exec-form image with --foo=bar command line argument:
```
$ docker run springcloudtask/billrun:execform --foo=bar &
$ docker exec -it $(docker ps -q) ps axw
PID TTY      STAT   TIME COMMAND
  1 ?        Ssl    0:00 java -Xms64m -Xmx64m -jar /app.jar --foo=bar
```
The --foo=bar command line argument is successfully passed to java process!

**Comment by sabbyanandan** [comment]
Hi, @AdVirMachina. As Christian pointed out, the incompatibility of Docker's SHELL spec when it comes to Spring Cloud Task + Spring Batch's core functionality of CLI Args doesn't work well together, unfortunately. 

Please have a look and retry any of the described options (via: EXEC or Boot). If you still have any other questions, leave a note here.

**Comment by AdVirMachina** [comment]
Unfortunately, my client refuses to change from SHELL. They view it as a security hole to have the containerized applications take command line parameters.

As to SCDF not being compatible with SHELL, I wound up passing environment variables to the execution and then loading them into the Environment before starting up Spring Batch in the container. This worked if the environment variables are part of the definition. One cannot do this in the GUI, but can through the REST API.

But, this still doesn't address the tracking issue. It seems to me that SCDF would be able to coordinate the executions by adding a context identifier on the command line when it does the submission to K8s. Much like I have to in the definition.

**Comment by sabbyanandan** [comment]
@AdVirMachina: What we suggested above leverage either the "exec" or "boot" as the options, and we are _not_ implying that you use the "shell". 

Did you try either of the choices?

**Comment by sabbyanandan** [comment]
Upon reading your response again, it appears your client is not willing to move out of Shell, so that would imply that "boot" or "exec" is not possible choices.

I'm curious as to what security issues do they foresee when using "boot" since everything is wrapped inside `SPRING_APPLICATION_JSON` representation. If you can describe specifically the concerns here, that'd help.

**Comment by AdVirMachina** [comment]
Their argument is based on how the templates for SCDF are delivered with Open Shift Enterprise. They feel that RedHat would have implemented the best practice for security. I don't know enough about SCDF in total to agree or disagree.

I am tasked with integrating SCDF Task submissions with the CA/7 Desktop Edition Scheduler system. I did all the development using EXEC. When it came time to integrate with their infrastructure OKD, I found out that SHELL was all they would allow. They have been absolutely adamant about it!

**Comment by tzolov** [comment]
Hi @AdVirMachina, 

`Spring Batch` (which optionally can be used from within `Spring Task`) requires command line arguments for configuring dynamic parameters. The `Docker SHELL` on the other  other hand doesn't allow passing command line arguments.  So as you can see those two will *never* work together and this has nothing to do with SCDF itself.

If you are to use only `Spring Task` (and not Spring Batch) then we've opened this https://github.com/spring-cloud/spring-cloud-dataflow/issues/3651 . When resolved would allow to running pure Spring Tasks (no Batch) in SCDF using the Docker SHELL mode. 

I am still not sure what are those best security practices referred above? 
The primary difference between EXEC and the SHELL is that EXEC runs the executables directly (e.g. `java -jar my.jar`) while SHELL runs them from within a bash shel (e.g. `bin\sh -c 'java -jar my.jar'`). 
If customer's requirements is to run the executable from within a bash shell this can be done with BOOT and even EXEC configuration.

**Comment by sabbyanandan** [comment]
As discussed in the last comment, outside of the "boot" and "exec" as options, we will enhance to accept CLI args as env-vars, which may address the "shell" requirement, as well. Closing this in favor of https://github.com/spring-cloud/spring-cloud-dataflow/issues/3651.

**Comment by AdVirMachina** [comment]
Tzolov,
   I think everyone is focusing on the wrong thing. I have solved my issues with EXEC running my program. 

  What I have issue with is the SCDF server is posting two executions in the database for one submission to K8S. I am interfacing SCDF with CA7 Desktop Edition and cannot use the immediately returned Execution ID from the REST API. The first one which is returned is for the Pod creation only. I have to wait for another one to appear in the list and do the correlation myself. My contention is the SCDF server should be doing this itself regardless of how the K8S profile executes my application.

**Comment by ilayaperumalg** [comment]
Reopening the issue based on the recent comment.

**Comment by umutcann** [comment]
Hi,

We have same issue, As detailed in #[3536](https://github.com/spring-cloud/spring-cloud-dataflow/issues/3536), 

Then, we tried to start our batch app on Openshift using entryPointStyle: exec mod.

We got error like below.

Error: container create failed: container_linux.go:345: starting container process caused "exec: \"--spring.datasource.username=dbname\": executable file not found in $PATH"

Our batch app was buily by S2I. So that It has no Dockerfile. 

So workaround which specified like above won't be work for us.

Do you have any comment about our issue?

**Comment by chrisjs** [comment]
@umutcann based on the error you posted:

```
Error: container create failed: container_linux.go:345: starting container process caused "exec: "--spring.datasource.username=dbname": executable file not found in $PATH"
```
would point to an error in your docker configuration. im not familiar with "S21" in regards to `Our batch app was buily by S2I. So that It has no Dockerfile.` -- https://www.openshift.com/blog/create-s2i-builder-image has explicit references to creating a Dockerfile.

Can you provide the Dockerfile in question?

**Comment by chrisjs** [comment]
issue has been open for awhile, if this is still an issue please re-open with a reproducible test case
